### PR TITLE
test(connlib): introduce routing table to `tunnel_test`

### PR DESCRIPTION
**Pull request by thomaseizinger**
Currently, `tunnel_test` uses a rather naive approach when dispatching `Transmit`s. In particular, it checks client, gateway and relay separately whether they "want" a certain packet. In a real network, these packets are routed based on their IP.

To mimic something similar, we introduce a `Host` abstraction that wraps each component: client, gateway and relay. Additionally, we introduce a `RoutingTable` where we can add and remove hosts. With these things in place, routing a `Transmit` is as easy as looking up the destination IP in the routing table and dispatching to the corresponding host.

Our hosts are type-safe: client, gateway and relay have different types. Thus, we abstract over them using a `HostId` in order to know, which host a certain message is for. Following these patches, we can easily introduce multiple gateways and relays to this test by simply making more entries in this routing table. This will increase the test coverage of connlib.

Lastly, this patch massively increases the performance of `tunnel_test`. It turns out that previously, we spent a lot of CPU cycles accessing "random" IPs from very large iterators. With this patch, we take a limited range of 100 IPs that we sample from, thus drastically increasing performance of this test. The configured 1000 testcases execute in 3s on my machine now (with opt-level 1 which is what we use in CI).

### CODE REVIEW

File: .github/codespellrc [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [codespell]
-skip = ./**/*.svg,./elixir/deps,./**/*.min.js,./kotlin/android/app/build,./kotlin/android/build,./e2e/pnpm-lock.yaml,./website/.next,./website/pnpm-lock.yaml,./rust/target,Cargo.lock,./website/docs/reference/api/*.mdx,./**/erl_crash.dump,./cover,./vendor,*.json,seeds.exs,./**/node_modules,./deps,./priv/static,./priv/plts,./**/priv/static,./.git,./_build,*.cast
+skip = ./**/*.svg,./elixir/deps,./**/*.min.js,./kotlin/android/app/build,./kotlin/android/build,./e2e/pnpm-lock.yaml,./website/.next,./website/pnpm-lock.yaml,./rust/target,Cargo.lock,./website/docs/reference/api/*.mdx,./**/erl_crash.dump,./cover,./vendor,*.json,seeds.exs,./**/node_modules,./deps,./priv/static,./priv/plts,./**/priv/static,./.git,./_build,*.cast,/rust/**/proptest-regressions/**/*.txt
```
Comment (thomaseizinger) [inline .github/codespellrc]:
```suggestion
skip = ./**/*.svg,./elixir/deps,./**/*.min.js,./kotlin/android/app/build,./kotlin/android/build,./e2e/pnpm-lock.yaml,./website/.next,./website/pnpm-lock.yaml,./rust/target,Cargo.lock,./website/docs/reference/api/*.mdx,./**/erl_crash.dump,./cover,./vendor,*.json,seeds.exs,./**/node_modules,./deps,./priv/static,./priv/plts,./**/priv/static,./.git,./_build,*.cast,./rust/**/proptest-regressions/**/*.txt
```

File: .github/codespellrc [hunk 2]
```diff
@@ -1,3 +1,3 @@
 [codespell]
-skip = ./**/*.svg,./elixir/deps,./**/*.min.js,./kotlin/android/app/build,./kotlin/android/build,./e2e/pnpm-lock.yaml,./website/.next,./website/pnpm-lock.yaml,./rust/target,Cargo.lock,./website/docs/reference/api/*.mdx,./**/erl_crash.dump,./cover,./vendor,*.json,seeds.exs,./**/node_modules,./deps,./priv/static,./priv/plts,./**/priv/static,./.git,./_build,*.cast
+skip = ./**/*.svg,./elixir/deps,./**/*.min.js,./kotlin/android/app/build,./kotlin/android/build,./e2e/pnpm-lock.yaml,./website/.next,./website/pnpm-lock.yaml,./rust/target,Cargo.lock,./website/docs/reference/api/*.mdx,./**/erl_crash.dump,./cover,./vendor,*.json,seeds.exs,./**/node_modules,./deps,./priv/static,./priv/plts,./**/priv/static,./.git,./_build,*.cast,./rust/**/proptest-regressions/**/*.txt
```
Comment (thomaseizinger) [inline .github/codespellrc]:
```suggestion
skip = ./**/*.svg,./elixir/deps,./**/*.min.js,./kotlin/android/app/build,./kotlin/android/build,./e2e/pnpm-lock.yaml,./website/.next,./website/pnpm-lock.yaml,./rust/target,Cargo.lock,./website/docs/reference/api/*.mdx,./**/erl_crash.dump,./cover,./vendor,*.json,seeds.exs,./**/node_modules,./deps,./priv/static,./priv/plts,./**/priv/static,./.git,./_build,*.cast,./**/proptest-regressions
```

File: rust/connlib/tunnel/src/tests/sim_net.rs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+}
+
+impl<S> Id for SimRelay<S> {
+    fn id(&self) -> ComponentId {
+        self.id.into()
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub(crate) enum ComponentId {
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/tests/sim_net.rs]:
```suggestion
pub(crate) enum HostId {
```

File: rust/connlib/tunnel/src/tests/reference.rs [hunk 4]
```diff
@@ -85,61 +88,85 @@ impl ReferenceStateMachine for ReferenceState {
...
-                    let c6_eq_g6 = c6.is_some_and(|c| g6.is_some_and(|g| c == g));
-                    let c4_eq_r4 = c4.is_some_and(|c| r4.is_some_and(|r| c == r));
-                    let c6_eq_r6 = c6.is_some_and(|c| r6.is_some_and(|r| c == r));
-                    let g4_eq_r4 = g4.is_some_and(|g| r4.is_some_and(|r| g == r));
-                    let g6_eq_r6 = g6.is_some_and(|g| r6.is_some_and(|r| g == r));
-
-                    !c4_eq_g4 && !c6_eq_g6 && !c4_eq_r4 && !c6_eq_r6 && !g4_eq_r4 && !g6_eq_r6
-                },
+                "client and gateway priv key must be different",
+                |(c, g, _, _, _, _, _, _, _)| c.inner().state.0 != g.inner().state,
```
Comment (conectado) [inline rust/connlib/tunnel/src/tests/reference.rs:148]: Do we need that restriction?

File: rust/connlib/tunnel/src/tests/sim_node.rs [hunk 5]
```diff
@@ -210,64 +124,54 @@ impl fmt::Debug for PrivateKey {
...
-
-        let mut subnets = cgnat_block.subnets_with_prefix(128);
-
-        subnets
-            .nth(idx.index(subnets.len()))
-            .unwrap()
-            .network_address()
-    })
+pub(crate) fn tunnel_ip6s() -> impl Iterator<Item = Ipv6Addr> {
+    Ipv6Network::new(Ipv6Addr::new(64_768, 8_225, 4_369, 0, 0, 0, 0, 0), 107)
```
Comment (conectado) [inline rust/connlib/tunnel/src/tests/sim_node.rs:171]: Can we write it in hex? `0xfd00::2021::1111` is easier to read

**Review by conectado — APPROVED** [review]
LGTM!

File: rust/connlib/tunnel/src/tests/reference.rs (hunk 4 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/tests/reference.rs:148]: Yeah otherwise both nodes have the same private key.

**Pull request merged**